### ISSUE TITLE
Lotus FEVM API Incompatibility with normal Ethereum mainnet behavior, causing issues with integration partners 

### ISSUE THREAD

**Issue by Schwartz10** [filecoin-project/lotus#12441]
### Checklist

- [X] This is **not** a security-related bug/issue. If it is, please follow please follow the [security policy](https://github.com/filecoin-project/lotus/security/policy).
- [X] I **have** searched on the [issue tracker](https://github.com/filecoin-project/lotus/issues) and the [lotus forum](https://github.com/filecoin-project/lotus/discussions), and there is no existing related issue or discussion.
- [X] I did not make any code changes to lotus.

### Lotus component

- [X] lotus Ethereum RPC
- [ ] lotus FVM - Lotus FVM interactions
- [ ] FEVM tooling
- [ ] Other

### Lotus Version

```text
Running the public glif nodes
```

### Repro Steps

Curl request to `eth_estimateGas` that fails with actor not found:

```
curl 'https://api.calibration.node.glif.io/rpc/v1' \
  -H 'Accept: application/json' \
  -H 'Accept-Language: en-US,en;q=0.9,es;q=0.8' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Cookie: __stripe_mid=778e36cb-7efb-46da-986b-8e8e1146f17e7602bc' \
  -H 'Origin: chrome-extension://nkbihfbeogaeaoehlefnkodbefgpgknn/' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: none' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36' \
  --data-raw '{"id":588633187,"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"from":"0xb1040d3ce131b3204e5229c80a8d5ae271b2ef09","to":"0x7fa33ed9fce1e9c620cef778c5601286cd68e027","data":"0x5c19a95c000000000000000000000000b1040d3ce131b3204e5229c80a8d5ae271b2ef09"}]}'
```

returns:

```
{"jsonrpc":"2.0","error":{"code":1,"message":"failed to estimate gas: ApplyWithGasOnState failed: call raw get actor: resolution lookup failed (t410fweca2phbggzsatssfheavdk24jy3f3yjxzll22i): resolve address t410fweca2phbggzsatssfheavdk24jy3f3yjxzll22i: actor not found"},"id":588633187}
```

This is an incompatibility with the same call on infura ethereum mainnet, which returns:

```
{"jsonrpc":"2.0","id":3525842942,"result":"0xbea9"}
```

### Describe the Bug

Integration partners call `eth_estimateGas` to estimate the gas of a given transaction. In some circumstances, a web app may estimate the gas of a transaction even if the account has no balance to pay for the transaction fee. Ethereum mainnet seems to ignore the `from` address' balance when calling `eth_estimateGas` . Subsequently, when the same user tries to submit the transaction with the same `from` address, the call fails with insufficient funds (or maybe in our case, actor not found).

In some cases, integrators have to patch filecoin specific functionality into their otherwise chain agnostic frontend 

### Tooling

```text
Curl
```

### Configuration Options

```text
This is the public glif api, i'm not sure its exact config options
```

**Comment by rvagg** [comment]
I've been trying to revive fevm-contract-tests for more regular runs against the latest lotus master and I've been running into this problem there too. I have 8 failures on the uniswap test suite, and all but 1 are variations of this bug.

```
  1) UniswapV3Factory
       #setOwner
         fails if caller is not owner:
     AssertionError: Expected transaction to be reverted, but other exception was thrown: ProviderError: failed to estimate gas: ApplyWithGasOnState failed: call raw get actor: resolution lookup failed (f410foji4f3nrgqcrctymqwe5gmppms2nvy6iqkoyaty): resolve address f410foji4f3nrgqcrctymqwe5gmppms2nvy6iqkoyaty: actor not found
```

So I guess we've introduced some more strict checking at some point in the pipeline here.

**Comment by jennijuju** [comment]
should try to repro w/ metamask and see what error response other evm chain returns (see the warning)
![image](https://github.com/user-attachments/assets/a0ed7f38-7cb4-440a-aa0b-3579a6902957)

**Comment by Schwartz10** [comment]
<img width="619" alt="image" src="https://github.com/user-attachments/assets/ac0ac037-760c-4b47-89ed-72f9dad4465c">
Ethereum mainnet

**Comment by rvagg** [comment]
hah, this call is working now on calibnet

```json
{"jsonrpc":"2.0","result":"0x925da8","id":588633187}
```

the actor has funds, or has an actorid now, or ... something?

if we modify the `from` manually to be some unknown address and we get the same error, e.g. changing the `9` on the end to an `8`, getting:

```json
{"jsonrpc":"2.0","error":{"code":1,"message":"failed to estimate gas: ApplyWithGasOnState failed: call raw get actor: resolution lookup failed (t410fweca2phbggzsatssfheavdk24jy3f3yixfykv3a): resolve address t410fweca2phbggzsatssfheavdk24jy3f3yixfykv3a: actor not found"},"id":588633187}
```

does geth / infura / whoever, still process it without the valid `from`? I don't have an eth endpoint to test this.

I'm assuming that we inserted an actor check in there at some point that shouldn't be there, but I'm honestly not sure what the 'correct' behaviour is. @Schwartz10 ?

**Comment by Abhay-2811** [comment]
Hey @rvagg  so I tried the calls and only the calls [Example: `eth_call` and `eth_estimateGas` ] that use `from` as a param throw this exact error. the error boils down to the issue with following piece of code:

[[// LookupIDAddress gets the ID address of this actor's `addr` stored in the `InitActor`.](https://github.com/filecoin-project/lotus/blob/dafc56c8051a06ecb7c60267e9f5852b8d856918/chain/state/statetree.go#L343-L361)](https://github.com/filecoin-project/lotus/blob/dafc56c8051a06ecb7c60267e9f5852b8d856918/chain/state/statetree.go#L343-L361)

A quick patch would be: checking for calls that don't require gas [ I only came across 2 so it's feasible ] and remove the `from` param from call. 

Do let me know your thoughts about this

**Comment by rvagg** [comment]
@akaladarshi has been looking into this one too, latest [thread on Slack](https://filecoinproject.slack.com/archives/CP50PPW2X/p1727969488125229).

@akaladarshi would you mind dropping in a summary of your findings so far? I thought your findings from go-ethereum were the most interesting bit—it basically shortcuts some of the internal checks and if we were to follow a similar path it would mean sending that into the fvm to have it avoid some of its checks. But I think we need to work out what all of those checks are first.

We really should build a solid, failing itest for this first that we can work on fixing. There are possibly more than one situation that we need to account for.

**Comment by rvagg** [comment]
Reopened because I don't think we actually fixed this, at least it needs testing.

**Comment by rvagg** [comment]
OK, thanks to @akaladarshi's prompting I went back and did a bit of a review of where this was at before it was closed, and also why it was closed.

The closing happened due to https://github.com/filecoin-project/fevm-contract-tests/pull/28, but I'm not quite sure how. I didn't put metadata in the commit message but maybe it was because I said "fix <this PR>" in the summary of that PR? But I actually refer to this issue and say that I introduced a minor hack in that PR (funding unfunded accounts) to make it not report this error, and that if we "fix" this issue then I could undo that. So it's not fixed because of that PR.

In the [Slack thread](https://filecoinproject.slack.com/archives/CP50PPW2X/p1727969488125229) already mentioned above, @akaladarshi did a look at what go-ethereum does here that's different to Lotus and it boils down to this: https://github.com/ethereum/go-ethereum/blob/16f64098b91d58541836abcb4f51fe1af7e0cdbc/core/state_transition.go#L145-L151 - `SkipNonceChecks` and `SkipFromEOACheck` are added to the message before execution so the EVM doesn't perform all its normal checks when it's doing a gas estimate. So in the case of the error we're seeing here, the account doesn't need to even exist to estimate the gas, it'll run as if it does just for the purpose of estimation. We don't have such a feature, we do all those checks and run the message as if the message could be submitted immediately.

Replicating this functionality would likely be quite a tricky job since so much depends on the state of the `from` actor, but maybe it's not too complicated for simple sends. I think it's legitimate to keep this issue open as a possible TODO to bridge the functionality/expectation gap with eth.

**_The workaround for this in the meantime is to do estimation from an actor that exists on chain._**

**Comment by Schwartz10** [comment]
Hey thanks @rvagg for looking back into this. I just want to note that:

"The workaround for this in the meantime is to do estimation from an actor that exists on chain."

Is not always a workable workaround, especially when interacting with sophisticated FEVM smart contracts that manage state, when the `from` actor that has no ID address is involved somehow in that state.  Happy to provide examples here (the one I've run into is [Tally governance](https://tally.xyz/) and the way their system/web app works), but we have issues with integration partner UIs because of this incompat. We'd be asking all normal EVM app developers to change their web app functionality to avoid errors for FEVM specific nuance, which is unlikely

**Comment by BigLep** [comment]
Agreed we can't be asking normal EVM app developers to change their web app functionality.  During Lotus maintainer standup, there was some discussion on how this could be solved.  @rvagg do you want to drop some notes while it's fresh (sorry - I didn't catch enough of the details.).  

@Schwartz10 : what's the impact to you or others until this gets addressed.  We definitely plan to handle this in 2025Q1, but we are currently prioritizing items that are needed for network upgrades first.  Depending on the impact of this issue, we will evaluate if there is a way to do it sooner.

**Comment by aarshkshah1992** [comment]
@BigLep @rvagg This is important even to get the SAFE multisig working well on Filecoin. SAFE is the best way we have right now to get multisigs to own/interact with smart contracts and so would be good to get this in for better contract management UX on Filecoin.

**Comment by rvagg** [comment]
@aarshkshah1992 is this the error that you see? I thought from Slack it was just what was seen when trying to replicate the same message as you were sending? Does SAFE need to be set up from a non-existing actor?

**Comment by rvagg** [comment]
Discussion today was that maybe we should either be rewriting the `from` to be the system actor, or just funding and setting up the actor if it doesn't exist before executing the message. @Stebalien should drop in other thoughts because neither of these sound ideal to me because in the case of doing it as the system actor won't we be stumbling over certain checks in builtin-actors in some cases? and in the case of making an actor appear on chain before we call the vm won't we be littering the blockstore with fake state just to get it running, thereby creating a way to bloat a public RPC node's blockstore before splitstore can GC it? Something like `SkipFromEOACheck` would be ideal, but it seems like a hard way of solving this because I bet there's a of checks that touch the message sending actor.

**Comment by Stebalien** [comment]
I looked into this a bit more and:

1. I don't think we should set it to the system actor now as that'll cause the called actor to see the wrong caller.
2. We don't use the "implicit" message type because "implicit" messages don't account for inclusion costs (because the message doesn't live on-chain).

I'd like to solve this by sending an implicit (i.e., free) message from the system actor to the EOA to create it (with zero funds) before checking gas. IMO, that's better than just ignoring the EOA because it'll force it to exist. This doesn't matter in Ethereum because, in Ethereum, all EOAs exist implicitly (with zero funds). In Filecoin, actors are created (sometimes explicitly) but their existence itself isn't implicit.

TL;DR: No changes to the FVM. Change `StateManager.callInternal` to add a `createSender` flag and, when set, apply an implicit message (`vmi.ApplyMessageImplicit`) from the "system" actor to the "from" actor with no funds attached to it.

**Comment by BigLep** [comment]
Given @Stebalien outlined a path forward (thank you!), if someone wants to take this up in the short term (anyone feel free to volunteer), Lotus volunteers will make sure to at least provide review.
